operator/internal/controller/pulumi/stack_controller.go

```diff
@@ -875,9 +873,72 @@ func newStackReconcilerSession(
 	}
 }
 
+// isSynced determines whether the stack is in sync with respect to the
+// specification. i.e. the current spec generation has been applied, the update
+// was successful, the latest commit has been applied, and (if resync is
+// enabled) has been resynced recently.
+func isSynced(stack *pulumiv1.Stack, currentCommit string, isStackMarkedToBeDeleted bool) bool {
+	if stack.Status.LastUpdate == nil {
+		return false
+	}
+
+	if stack.Status.LastUpdate.Generation != stack.Generation {
+		return false
+	}
+
+	if stack.Status.LastUpdate.State == shared.SucceededStackStateMessage {
+		if isStackMarkedToBeDeleted {
+			return true
+		}
+		if stack.Status.LastUpdate.LastSuccessfulCommit != currentCommit {
+			return false
+		}
+		if !stack.Spec.ContinueResyncOnCommitMatch {
+			return true
+		}
+		return time.Since(stack.Status.LastUpdate.LastResyncTime.Time) < resyncFreq(stack)
+	}
+
+	if stack.Status.LastUpdate.State == shared.FailedStackStateMessage {
+		return time.Since(stack.Status.LastUpdate.LastResyncTime.Time) < cooldown(stack)
+	}
+
+	// We should never get here; if we do the Update is in an unknown state so
+	// don't trigger work for it.
+	return true
+}
+
+// cooldown returns the amount of time to wait before a failed Update should be
+// retried. We start with a 5-minute cooldown and double that for each failed
+// attempt, up to a max of 24 hours. Failed Updates are considered synced while
+// inside this cooldown period. A zero-value duration is returned if the update
+// succeeded.
+func cooldown(stack *pulumiv1.Stack) time.Duration {
+	cooldown := time.Duration(0)
+	if stack.Status.LastUpdate == nil {
+		return cooldown
+	}
+	if stack.Status.LastUpdate.State == shared.FailedStackStateMessage {
+		cooldown = 5 * time.Minute
+		cooldown *= time.Duration(math.Exp2(float64(stack.Status.LastUpdate.Failures)))
+		cooldown = min(24*time.Hour, cooldown)
```

Comment:
This seems not aggressive enough to cover short-term transient errors like a network connection error.

Reply:
Lowered the initial backoff to 1 minute.